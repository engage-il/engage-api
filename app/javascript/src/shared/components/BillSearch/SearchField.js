// @flow
import * as React from 'react'
import styled from 'react-emotion'
import { Link } from '../Link'
import { colors, mixins } from 'shared/styles'

type Props = {
  value: string,
  onChange: (string) => void
}

type State = {
  isFocused: boolean
}

export class SearchField extends React.Component<*, Props, State> {
  state = {
    isFocused: false
  }

  // events
  inputDidChange = (event: { target: { value: string } }) => {
    const { onChange } = this.props
    onChange(event.target.value)
  }

  inputDidChangeFocus = (isFocused: boolean) => () => {
    this.setState({ isFocused })
  }

  // lifecycle
  render () {
    const { value } = this.props
    const { isFocused } = this.state

    return (
      <Container>
        <Field isFocused={isFocused}>
          <Input
            type='text'
            name='search-field'
            value={value}
            placeholder={`health care, HB2364`}
            onChange={this.inputDidChange}
            onFocus={this.inputDidChangeFocus(true)}
            onBlur={this.inputDidChangeFocus(false)}
          />
          <Button
            onClick={() => {}}
            children='Search'
          />
        </Field>
      </Container>
    )
  }
}

const Container = styled.div`
  ${mixins.flexRow};

  justify-content: center;
  align-items: center;
  height: 190px;
  background-color: ${colors.primary};
`

const Field = styled.div`
  ${mixins.flexRow};
  ${mixins.pageWidth};
  ${mixins.pageMargin};

  flex: 1;
  height: 70px;
  border: solid ${colors.white};
  border-width: ${({ isFocused }: State) => isFocused ? 2 : 1}px;
  border-radius: 10px;
  overflow: hidden;
  transform: perspective(200px) translateZ(${({ isFocused }: State) => isFocused ? 1 : 0}px);
  font-size: 24px;
  color: ${colors.white};
  transition: border-width 0.15s, transform 0.15s;
`

const Input = styled.input`
  flex: 1;
  padding: 0 30px;
  color: inherit;
  background-color: transparent;
  font-size: inherit;

  ::placeholder {
    color: ${colors.white};
    opacity: 0.5;
  }
`

const Button = styled(Link)`
  ${mixins.flexRow};
  ${mixins.flexCenter};
  ${mixins.fonts.bold};

  width: 215px;
  color: ${colors.primary};
  background-color: ${colors.white};
  text-transform: uppercase;
`
