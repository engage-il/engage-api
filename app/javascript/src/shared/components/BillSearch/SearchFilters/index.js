// @flow
export { SearchFilters } from './SearchFilters'
