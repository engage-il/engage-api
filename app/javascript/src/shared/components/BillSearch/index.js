// @flow
export { BillSearch, initialVariables } from './BillSearch'
