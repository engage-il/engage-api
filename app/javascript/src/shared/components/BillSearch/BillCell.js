// @flow
import * as React from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import { css } from 'glamor'
import { formatDate } from 'shared/date'
import type { Bill } from 'shared/types'
import { Button } from 'shared/components'
import { stylesheet, colors, mixins } from 'shared/styles'

type Props = {
  bill: Bill,
  className?: string
}

let BillCell = function BillCell ({
  bill
}: Props) {
  const formattedDate = formatDate(bill.hearing.date, 'DD/MM/YYYY')

  return (
    <div {...css(rules.container)}>
      <div {...rules.info}>
        <div {...rules.header}>
          <div {...rules.document}>
            <h3>{bill.title}</h3>
            <span>{bill.documentNumber}</span>
          </div>
          <p>{formattedDate}</p>
        </div>
        {bill.summary && <p {...rules.summary}>{bill.summary}</p>}
      </div>
      <div {...rules.actions}>
        <Button
          styles={rules.button}
          to={bill.witnessSlipUrl}
          children='Take Action'
        />
        <Button
          styles={rules.button}
          to={`/bill/${bill.id}`}
          isSecondary
          children='More Info'
        />
      </div>
    </div>
  )
}

BillCell = createFragmentContainer(BillCell, graphql`
  fragment BillCell_bill on Bill {
    id
    documentNumber
    title
    summary
    witnessSlipUrl
    detailsUrl
    fullTextUrl
    hearing {
      date
    }
  }
`)

const rules = stylesheet({
  container: {
    ...mixins.shadows.low,
    ...mixins.borders.low(),
    display: 'flex',
    padding: 15,
    marginBottom: 15,
    backgroundColor: colors.neutral,
    '&:last-child': {
      marginBottom: 0
    },
    ...mixins.mobile.glam({
      flexDirection: 'column'
    })
  },
  info: {
    ...mixins.borders.low(['right']),
    flex: 1,
    paddingRight: 15,
    ...mixins.mobile.glam({
      ...mixins.borders.low(['bottom']),
      paddingRight: 0,
      paddingBottom: 15,
      marginBottom: 15,
      borderRight: 'none'
    })
  },
  header: {
    display: 'flex',
    flexDirection: 'column'
  },
  document: {
    marginBottom: 5,
    '> *': {
      display: 'inline-block'
    },
    '> h3': {
      ...mixins.fonts.bold,
      marginRight: 10
    }
  },
  actions: {
    display: 'flex',
    flexDirection: 'column',
    paddingLeft: 15,
    ...mixins.mobile.glam({
      flexDirection: 'row',
      paddingLeft: 0
    })
  },
  button: {
    width: 200,
    marginBottom: 10,
    ':last-child': {
      marginBottom: 0
    },
    ...mixins.mobile.glam({
      flex: 1,
      width: 'auto',
      marginBottom: 0,
      marginRight: 10,
      ':last-child': {
        marginRight: 0
      }
    })
  },
  summary: {
    ...mixins.borders.low(['top']),
    marginTop: 10,
    paddingTop: 10
  }
})

export { BillCell }
