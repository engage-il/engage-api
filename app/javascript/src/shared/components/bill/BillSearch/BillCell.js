// @flow
import * as React from 'react'
import styled from 'react-emotion'
import { createFragmentContainer, graphql } from 'react-relay'
import { BillTitle } from '../BillTitle'
import { BillStatus } from '../BillStatus'
import { Button } from '@/components'
import { mixins } from '@/styles'
import type { Bill } from '@/types'

type Props = {
  bill: Bill,
  className?: string
}

let BillCell = function BillCell ({
  bill
}: Props) {
  return (
    <div>
      <Cell>
        <BillTitle bill={bill} />
        <BillStatus bill={bill} />
        <p>{bill.summary}</p>
        <Button to={`/bill/${bill.id}`} isSmall children='More Info' />
      </Cell>
    </div>
  )
}

BillCell = createFragmentContainer(BillCell, graphql`
  fragment BillCell_bill on Bill {
    id
    summary
    hearing {
      date
    }
    ...BillTitle_bill
    ...BillStatus_bill
  }
`)

const Cell = styled.div`
  ${mixins.flexColumn};

  align-items: flex-start;
  max-width: 960px;

  > p {
    margin-bottom: 15px;
  }
`

export { BillCell }
