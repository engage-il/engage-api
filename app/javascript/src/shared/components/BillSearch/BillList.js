// @flow
import * as React from 'react'
import { createPaginationContainer, graphql } from 'react-relay'
import type { RelayPaginationProp } from 'react-relay'
import { withRouter } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import styled from 'react-emotion'
import { BillCell } from './BillCell'
import { TranslateAndFade, Button } from '@/components'
import { session } from '@/storage'
import { withLoadMoreArgs } from '@/relay'
import { colors, mixins } from '@/styles'
import type { Viewer } from '@/types'

type Props = {
  viewer: Viewer,
  isAnimated: boolean,
  pageSize?: number,
  relay: RelayPaginationProp
} & ContextRouter

type State = {
  disablesAnimation: boolean
}

function formatCount ({ bills }: Viewer) {
  return `Found ${bills.count} result${bills.count === 1 ? '' : 's'}.`
}

let BillList = class BillList extends React.Component<*, Props, State> {
  state = {
    disablesAnimation: this.isFromPop
  }

  // accessors
  get isFromPop () {
    return this.props.history.action === 'POP'
  }

  // events
  didClickLoadMore = () => {
    const { relay, pageSize } = this.props
    if (!relay.hasMore() || relay.isLoading()) {
      return
    }

    pageSize && relay.loadMore(pageSize, (error: ?Error) => {
      if (error) {
        console.error(`error loading next page: ${error.toString()}`)
      }
    })
  }

  // lifecycle
  componentDidMount () {
    if (this.isFromPop) {
      this.setState({ disablesAnimation: false })
    }
  }

  componentWillUnmount () {
    const { viewer } = this.props
    viewer && session.set('last-search-count', `${viewer.bills.edges.length}`)
  }

  render () {
    const { viewer, isAnimated, pageSize, relay } = this.props
    const { disablesAnimation } = this.state

    return (
      <Bills>
        <h5>{formatCount(viewer)}</h5>
        <TranslateAndFade
          component={List}
          disable={!isAnimated || disablesAnimation}
        >
          {viewer.bills.edges.map(({ node }) => (
            <div key={node.id}>
              <BillCell bill={node} />
            </div>
          ))}
        </TranslateAndFade>
        {pageSize && relay.hasMore() && (
          <ActionButton
            isSecondary
            onClick={this.didClickLoadMore}
            children='Load More'
          />
        )}
      </Bills>
    )
  }
}

BillList = createPaginationContainer(withRouter(BillList),
  graphql`
    fragment BillList_viewer on Viewer {
      bills(
        filter: $filter,
        first: $count,
        after: $cursor
      ) @connection(key: "BillList_bills") {
        count
        pageInfo {
          hasNextPage
          endCursor
        }
        edges {
          node {
            id
            ...BillCell_bill
          }
        }
      }
    }
  `,
  withLoadMoreArgs({
    getConnectionFromProps (props) {
      return props.viewer && props.viewer.bills
    },
    query: graphql`
      query BillListQuery(
        $filter: BillsSearchFilter!,
        $count: Int!,
        $cursor: String!
      ) {
        viewer {
          ...BillList_viewer
        }
      }
    `
  })
)

const spacing = 50

const Bills = styled.div`
  ${mixins.flexColumn};

  > h5 {
    margin-bottom: ${spacing}px;
  }
`

const List = styled.div`
  > * {
    margin-bottom: ${spacing}px;
    padding-bottom: ${spacing}px;
    border-bottom: 1px solid ${colors.gray4};
  }
`

const ActionButton = styled(Button)`
  align-self: center;
  margin-bottom: 90px;
`

export { BillList }
