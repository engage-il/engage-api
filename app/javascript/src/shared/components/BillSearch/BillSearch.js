// @flow
import * as React from 'react'
import { createRefetchContainer, graphql } from 'react-relay'
import type { RelayRefetchProp } from 'react-relay'
import styled from 'react-emotion'
import { throttle } from 'lodash'
import { SearchField } from './SearchField'
import { BillList } from './BillList'
import { LoadingIndicator } from '../LoadingIndicator'
import type { Viewer, SearchParams } from '@/types'
import { mixins } from '@/styles'

type Props = {
  viewer: ?Viewer,
  pageSize?: number,
  params?: SearchParams,
  onFilter?: (SearchParams) => void,
  relay: RelayRefetchProp
}

type State = {
  query: string,
  disablesAnimation: boolean
}

let BillSearch = class BillSearch extends React.Component<*, Props, State> {
  state = {
    query: '',
    disablesAnimation: false,
    ...this.props.params
  }

  // actions
  filterBillsForQuery = throttle((query: string) => {
    const { onFilter, relay } = this.props

    const filter = { query }
    onFilter && onFilter(filter)

    this.setState({ disablesAnimation: true })
    relay.refetch({ filter }, null, (error: ?Error) => {
      if (error) {
        console.error(`error updaing query: ${error.toString()}`)
      }

      // completion comes back before render
      global.requestAnimationFrame(() => {
        this.setState({ disablesAnimation: false })
      })
    })
  }, 300)

  // events
  searchFieldDidChange = (query: string) => {
    this.setState({ query })
    this.filterBillsForQuery(query)
  }

  // lifecycle
  render () {
    const {
      viewer,
      pageSize
    } = this.props

    const {
      query,
      disablesAnimation
    } = this.state

    return (
      <Search>
        <SearchField
          value={query}
          onChange={this.searchFieldDidChange}
        />
        <Bills>
          <LoadingIndicator
            isLoading={!viewer}
          />
          {viewer && (
            <BillList
              viewer={viewer}
              pageSize={pageSize}
              isAnimated={!disablesAnimation}
            />
          )}
        </Bills>
      </Search>
    )
  }
}

BillSearch = createRefetchContainer(BillSearch,
  graphql`
    fragment BillSearch_viewer on Viewer {
      bills(
        filter: $filter,
        first: $count,
        after: $cursor
      ) @connection(key: "BillSearch_bills") {
        edges {
          node {
            id
          }
        }
      }
      ...BillList_viewer
    }
  `,
  graphql`
    query BillSearchQuery(
      $filter: BillsSearchFilter!,
      $count: Int!,
      $cursor: String!
    ) {
      viewer {
        ...BillSearch_viewer
      }
    }
  `
)

const Search = styled.div`
  ${mixins.flexColumn};
`

const Bills = styled.div`
  ${mixins.pageWidth};

  position: relative;
  align-self: center;
  margin-top: 40px;
`

export { BillSearch }
