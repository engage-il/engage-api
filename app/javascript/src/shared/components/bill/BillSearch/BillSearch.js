// @flow
import * as React from 'react'
import { createRefetchContainer, graphql } from 'react-relay'
import type { RelayRefetchProps } from 'react-relay'
import styled from 'react-emotion'
import { throttle } from 'lodash'
import { SearchField } from './SearchField'
import { SearchFilters } from './SearchFilters'
import { BillList } from './BillList'
import { BILL_SPACING, BILL_SPACING_MOBILE } from './constants'
import { Loading } from '@/components'
import { mixins } from '@/styles'
import type { Viewer, SearchParams } from '@/types'

type Props = {
  viewer: ?Viewer,
  pageSize?: number,
  params?: SearchParams,
  onChange?: (SearchParams) => void,
} & RelayRefetchProps

type State = {
  params: SearchParams,
  disablesAnimation: boolean
}

let BillSearch = class BillSearch extends React.Component<Props, State> {
  state = {
    params: {
      query: '',
      ...(this.props.params || {})
    },
    disablesAnimation: false
  }

  // actions
  filterBills = throttle((params: SearchParams) => {
    const { onChange, relay } = this.props

    // callback the change handler
    onChange && onChange(params)

    // fetch the bills with updated params
    this.setState({ disablesAnimation: true })

    // $FlowFixMe: variables failiing to match object
    relay.refetch({ params }, null, (error: ?Error) => {
      if (error) {
        console.error(`error updaing query: ${error.toString()}`)
      }

      // HACK: completion comes back before render
      global.requestAnimationFrame(() => {
        this.setState({ disablesAnimation: false })
      })
    })
  }, 300)

  // events
  didChangeParams = (params) => {
    this.setState({ params })
    this.filterBills(params)
  }

  // lifecycle
  render () {
    const {
      viewer,
      pageSize
    } = this.props

    const {
      params,
      disablesAnimation
    } = this.state

    return (
      <Search>
        <SearchField
          params={params}
          onChange={this.didChangeParams}
        />
        <Body>
          <SearchFilters
            params={params}
            onChange={this.didChangeParams}
          />
          <Bills>
            <Loading
              isLoading={!viewer}
            />
            {viewer && (
              <BillList
                viewer={viewer}
                pageSize={pageSize}
                isAnimated={!disablesAnimation}
              />
            )}
          </Bills>
        </Body>
      </Search>
    )
  }
}

BillSearch = createRefetchContainer(BillSearch,
  graphql`
    fragment BillSearch_viewer on Viewer {
      bills(
        params: $params,
        first: $count,
        after: $cursor
      ) @connection(key: "BillSearch_bills") {
        edges {
          node {
            id
          }
        }
      }
      ...BillList_viewer
    }
  `,
  graphql`
    query BillSearchQuery(
      $params: BillsSearchParams!,
      $count: Int!,
      $cursor: String!
    ) {
      viewer {
        ...BillSearch_viewer
      }
    }
  `
)

const Search = styled.div`
  ${mixins.flexColumn};
`

const Body = styled.div`
  ${mixins.pageContent};
  align-self: center;
`

const Bills = styled.div`
  position: relative;
  padding-top: ${BILL_SPACING}px;

  ${mixins.mobile`
    padding-top: ${BILL_SPACING_MOBILE}px;
  `}
`

export { BillSearch }
