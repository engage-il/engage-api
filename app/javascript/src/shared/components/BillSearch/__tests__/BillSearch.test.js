/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { defaultsDeep } from 'lodash'
import { BillSearch } from '../BillSearch'

const { anything } = expect

// subject
let subject

const defaults = {
  viewer: 'test-viewer',
  pageSize: 'test-size',
  onFilter: jest.fn(),
  relay: {
    refetch: jest.fn()
  }
}

function loadSubject (props = {}) {
  subject = shallow(<BillSearch {...defaultsDeep(props, defaults)} />)
}

// specs
afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('shows the search view', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })

  it('shows the search view with default params', () => {
    loadSubject({
      params: {
        query: 'test-query'
      }
    })

    expect(subject).toMatchSnapshot()
  })

  it('shows the loading indicator when loading', () => {
    loadSubject({
      viewer: null
    })

    expect(subject).toMatchSnapshot()
  })
})

describe('#filterBillsForQuery', () => {
  it('disables animations', () => {
    loadSubject()
    subject.instance().filterBillsForQuery('foo')
    expect(subject).toHaveState('disablesAnimation', true)
  })

  it('reteches the bills', () => {
    loadSubject()
    subject.instance().filterBillsForQuery('foo')
    expect(defaults.relay.refetch).toHaveBeenCalledWith({ filter: { query: 'foo' } }, null, anything())
  })

  it('notifies the handler', () => {
    loadSubject()
    subject.instance().filterBillsForQuery('foo')
    expect(defaults.onFilter).toHaveBeenCalledWith({ query: 'foo' })
  })
})

describe('entering a query', () => {
  it('filters bills by the query', () => {
    loadSubject()
    const instance = subject.instance()
    instance.filterBillsForQuery = jest.fn()

    instance.searchFieldDidChange('foo')
    expect(subject).toHaveState('query', 'foo')
    expect(instance.filterBillsForQuery).toHaveBeenCalled()
  })
})
