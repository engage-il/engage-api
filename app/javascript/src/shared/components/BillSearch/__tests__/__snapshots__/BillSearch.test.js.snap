// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render shows the loading indicator when loading 1`] = `
<Styled(div)>
  <SearchField
    onChange={[Function]}
    value=""
  />
  <Styled(div)>
    <LoadingIndicator
      isLoading={true}
    />
  </Styled(div)>
</Styled(div)>
`;

exports[`#render shows the search view 1`] = `
<Styled(div)>
  <SearchField
    onChange={[Function]}
    value=""
  />
  <Styled(div)>
    <LoadingIndicator
      isLoading={false}
    />
    <BillList
      isAnimated={true}
      pageSize="test-size"
      viewer="test-viewer"
    />
  </Styled(div)>
</Styled(div)>
`;

exports[`#render shows the search view with default params 1`] = `
<Styled(div)>
  <SearchField
    onChange={[Function]}
    value="test-query"
  />
  <Styled(div)>
    <LoadingIndicator
      isLoading={false}
    />
    <BillList
      isAnimated={true}
      pageSize="test-size"
      viewer="test-viewer"
    />
  </Styled(div)>
</Styled(div)>
`;
